### PR TITLE
[New features] Add hash routing, softplus score functions, and clamped SwiGLU

### PR DESCRIPTION
**Pull request by adam-xiaoyao**
### PR Category
Operator Mechanism

### PR Types
New features

### Description
本 PR 在 MoE 训练栈上新增/重写 3 个功能：

**1. Hash MoE Routing**：重写 hash 路由实现，使用 `tid2eid` lookup buffer（可从 checkpoint 加载），激活范围改为前 `moe_n_hash_layers` 个 MoE 层（0-indexed leading layers），权重来自 gate logits 经 score function gather，hash 层禁用 expert bias。新增 `moe_n_hash_layers` 和 `actual_vocab_size` 配置项。

**2. SoftPlus / SqrtSoftPlus 评分函数**：为 router 评分函数新增 `sftplus`（softplus，非负无界）选项；hash 路由专属 `sqrtsoftplus`（sqrt(softplus)）选项。

**3. Clamped SwiGLU**：新增 `clamped_swiglu` / `clamped_weighted_swiglu` 融合算子及对应反向，`clamp_value` 接入 `weighted_bias_swiglu_impl` 调用链，支持数值稳定性需求的训练配方。

<br>
<b>注：</b>swiglu提供了fuse和非fuse两种实现，如果要快速跑通非fuse的实现，请同时关掉以下flag才能进入非fuse分支

```yaml
using_sonic_moe: false
moe_use_fusion_node: false
moe_expert_fusion: false
moe_grouped_gemm: false
```

### CODE REVIEW

File: src/paddlefleet/transformer/moe/moe_router.py [hunk 1]
```diff
@@ -639,6 +646,69 @@ def _topk_noaux_tc(
...
+                of ``logits``. Shape [num_tokens], dtype int64.
+
+        Returns:
+            top_gate (paddle.Tensor): Per-token weights for the selected experts,
+                shape [num_tokens, topk]. Already normalized for non-softmax
+                score functions.
+            top_idx (paddle.Tensor): Selected expert indices, shape
+                [num_tokens, topk], dtype int64.
+        """
+        assert self.tid2eid is not None, (
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_router.py]: OK, 我修复

File: src/paddlefleet/transformer/transformer_config.py [hunk 2]
```diff
@@ -462,6 +463,16 @@ class TransformerConfig(ModelParallelConfig):
     moe_router_force_load_balancing: bool = False
     """Force load balancing with random logits for MoE router."""
 
+    moe_n_hash_layers: int = 0
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/transformer_config.py:455]: 当前 yaml_arguments.py 是基于 OmegaConf 的通用 flatten 机制，没有任何字段需要单独写解析逻辑。新增 moe_n_hash_layers / actual_vocab_size 只要存在于 TransformerConfig 的 dataclass 字段中（本 PR 已加），就会通过 core_transformer_config_from_args 中的 dataclasses.fields() 迭代自动从 YAML 加载。可参考同 PR 中其它新增字段（如 moe_router_score_function 相关）也未在 yaml_arguments.py 中单独处理。

[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/transformer_config.py:455]: 当前 yaml_arguments.py 是基于 OmegaConf 的通用 flatten 机制，没有任何字段需要单独写解析逻辑。新增 moe_n_hash_layers / actual_vocab_size 只要存在于 TransformerConfig 的 dataclass 字段中（本 PR 已加），就会通过 core_transformer_config_from_args 中的 dataclasses.fields() 迭代自动从 YAML 加载。可参考同 PR 中其它新增字段（如 moe_router_score_function 相关）也未在 yaml_arguments.py 中单独处理。

File: src/paddlefleet/transformer/moe/moe_router.py [hunk 3]
```diff
@@ -639,6 +646,69 @@ def _topk_noaux_tc(
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_router.py:741]: 暂时没必要， 这个和是否多卡关系不大

File: src/paddlefleet/fusions/fused_bias_swiglu.py [hunk 4]
```diff
@@ -217,21 +308,30 @@ def backward(ctx, grad_output):
...
-        tmp, wgrad = weighted_swiglu_back(grad_output, input, weights)
-        return tmp, wgrad, None
+        clamp_value = ctx.clamp_value
+        if clamp_value is not None:
+            tmp, wgrad = clamped_weighted_swiglu_back(
+                grad_output, input, weights, clamp_value
+            )
+        else:
+            tmp, wgrad = weighted_swiglu_back(grad_output, input, weights)
+        return tmp, wgrad
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/fusions/fused_bias_swiglu.py:337]: 已按review意见改了

File: src/paddlefleet/transformer/transformer_config.py [hunk 5]
```diff
@@ -456,12 +454,25 @@ class TransformerConfig(ModelParallelConfig):
...
+    moe_grouped_gemm: bool = False
+    """Whether to use grouped gemm."""
+
     router_z_loss_coef: float = None
     """Scaling coefficient for z-loss. Default is None."""
 
     moe_router_force_load_balancing: bool = False
     """Force load balancing with random logits for MoE router."""
 
+    moe_n_hash_layers: int = 0
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/transformer_config.py:455]: 不需要同步， yaml_argument.py里直接循环读取的。
